vault-plugins/flant_iam_auth/backend/backend.go

```diff
@@ -348,3 +351,196 @@ const (
+	vaultClient, err := b.accessVaultController.APIClient()
+	if err != nil {
+		return "", nil, fmt.Errorf("internal error accessing vault client: %w", err)
+	}
+
+	entityApi := entity_api.NewIdentityAPI(vaultClient, logger.Named("LoginIdentityApi")).EntityApi()
+	ent, err := entityApi.GetByID(req.EntityID)
+	if err != nil {
+		return "", nil, fmt.Errorf("finding vault entity by id: %w", err)
+	}
+
+	name, ok := ent["name"]
+	if !ok {
+		return "", nil, fmt.Errorf("field 'name' in vault entity is ommited")
+	}
+
+	nameStr, ok := name.(string)
+	if !ok {
+		return "", nil, fmt.Errorf("field 'name' should be string")
+	}
+	logger.Debug(fmt.Sprintf("catch name of vault entity: %s", nameStr))
+
+	txn := b.storage.Txn(false)
+	defer txn.Abort()
+
+	iamEntity, err := repo.NewEntityRepo(txn).GetByName(nameStr)
+	if err != nil {
+		return "", nil, fmt.Errorf("finding iam_entity by name:%w", err)
+	}
+	logger.Debug(fmt.Sprintf("catch multipass owner UUID: %s, try to find user", iamEntity.UserId))
+
+	user, err := iam_repo.NewUserRepository(txn).GetByID(iamEntity.UserId)
+	if err != nil && !errors.Is(err, iam.ErrNotFound) {
+		return "", nil, fmt.Errorf("finding user by id:%w", err)
+	}
+	if err == nil {
+		logger.Debug(fmt.Sprintf("found user UUID: %s", user.UUID))
+		return iam.UserType, user, nil
+	} else {
+		logger.Debug("Not found user, try to find service_account")
+		sa, err := iam_repo.NewServiceAccountRepository(txn).GetByID(iamEntity.UUID)
+		if err != nil && !errors.Is(err, iam.ErrNotFound) {
+			return "", nil, fmt.Errorf("finding service_account by id:%w", err)
+		}
+		if errors.Is(err, iam.ErrNotFound) {
+			logger.Debug("Not found neither user nor service_account")
+			return "", nil, err
+		}
+		logger.Debug(fmt.Sprintf("found service_account UUID: %s", sa.UUID))
+		return iam.ServiceAccountType, sa, nil
+	}
+}
+
+// availableTenantsAndProjectsByEntityIDOwner returns sets of tenants and projects available for EntityIDOwner
+func (b *flantIamAuthBackend) availableTenantsAndProjectsByEntityIDOwner(ctx context.Context,
+	req *logical.Request) (map[iam.TenantUUID]struct{}, map[iam.ProjectUUID]struct{}, error) {
+	subjectType, subject, err := b.revealEntityIDOwner(ctx, req)
+	if errors.Is(err, iam.ErrNotFound) {
+		return map[iam.TenantUUID]struct{}{}, map[iam.ProjectUUID]struct{}{}, nil
+	}
+	if err != nil {
+		return nil, nil, err
+	}
+	switch subjectType {
+	case iam.UserType:
+		{
+			user, ok := subject.(*iam.User)
+			if !ok {
+				return nil, nil, fmt.Errorf("can't cast, need *model.User, got: %T", subject)
+			}
+			txn := b.storage.Txn(false)
+			defer txn.Abort()
+			groups, err := iam_repo.NewGroupRepository(txn).FindAllParentGroupsForUserUUID(user.UUID)
+			gs := make([]iam.GroupUUID, 0, len(groups))
+			for uuid := range groups {
+				gs = append(gs, uuid)
+			}
+			if err != nil {
+				return nil, nil, fmt.Errorf("collecting tenants, get groups: %w", err)
+			}
+			//  TODO Here easy to have two or three paralleled goroutines
```

Comment:
I'd say it's arguable whether concurrency is worthwhile here

Reply:
it looks like this method will be run very often, it is just a reminder what to do if in profiling we really  will see the problem with this method